packages/core/src/index.ts

```diff
@@ -324,55 +351,116 @@ function needsToRecompute(target: Computed | Effect): boolean {
-		if (rollbackNode !== undefined) {
-			node._rollbackNode = rollbackNode;
-		}
+	const size = target._sources.length;
+
+	for (let i = 0; i < size; i++) {
+		const node = target._sources[i]!;
+		const back = node._source._node;
+
+		if (back !== undefined) node._rollbackNode = back;
+
 		node._source._node = node;
 		node._version = -1;
 	}
+
+	/**
+	 * If there are previous sources, e.g: [A, B, C]
+	 *                                      ^ Index is here at 0
+	 * As new deps/sources are added, we'll move or keep the current node at index,
+	 * If sources are unchanged, the resulting array's items won't have changed,
+	 * and the index will be equal to the array's sources length.
+	 */
+	target._index = 0;
 }
 
 function cleanupSources(target: Computed | Effect) {
-	// At this point target._sources is a mishmash of current & former dependencies.
-	// The current dependencies are also in a reverse order of use.
-	// Therefore build a new, reverted list of dependencies containing only the current
-	// dependencies in a proper order of use.
-	// Drop former dependencies from the list and unsubscribe from their change notifications.
-
-	let node = target._sources;
-	let sources = undefined;
-	while (node !== undefined) {
-		const next = node._nextSource;
-		if (node._version === -1) {
-			node._source._unsubscribe(node);
-			node._nextSource = undefined;
-		} else {
-			if (sources !== undefined) {
-				sources._prevSource = node;
-			}
-			node._prevSource = undefined;
-			node._nextSource = sources;
-			sources = node;
-		}
-
+	const size = target._sources.length;
+	const stop = target._index;
+
+	/**
+	 * Case 1: Best case scenario - source are the same as before. Example:
+	 *
+	 * Prev. sources: [A, B, C, D]
+	 * Next. sources: [A, B, C, D]...] (unchanged)
+	 *                             ↓
+	 *                          _index is length of array (no nodes to clean up)
+	 *
+	 * We only need to rollback to the previous node.
+	 */
+	for (let i = 0; i < stop; i++) {
+		const node = target._sources[i]!;
 		node._source._node = node._rollbackNode;
-		if (node._rollbackNode !== undefined) {
-			node._rollbackNode = undefined;
-		}
-		node = next;
 	}
-	target._sources = sources;
+	/**
+	 * Case 2: Worst case, example:
+	 *
+	 * Prev. sources: [A, B, C, D]
+	 * Next. sources: [A, B, E, F, G, D, C]
+	 *                       │  │  │  ↓
+	 *                       │  │  ↓  _index is here [D, C] are unused sources
+	 *                       │  ↓  3) G replaced prev node C, C moved to end of array -> [A, B, E, F, G, D, C]
+	 *                       ↓  2) F replaced prev node D, D pushed to end of array -> [A, B, E, F, C, D]
+	 *                      1) E replaced prev node C, C moved to end of array -> [A, B, E, D, C]
+	 *
+	 * Unsubscribe from nodes [D, C]
+	 */
+	for (let i = stop; i < size; i++) {
+		const node = target._sources[i]!;
+		node._source._node = node._rollbackNode;
+		node._source._unsubscribe(node);
```

Comment:
Wondering if we need to keep the `if (node._version === -1) {}` check before unsubscribing here. Afaik those were done to avoid unsubscribing from nodes that move in the sources array but still get reused. Might be something missing in our test suite